python_ta/checkers/unnecessary_assignment_checker.py

```diff
@@ -0,0 +1,99 @@
+        # >1 : assigned to before use.
+        # and the second element the last assignment node,
+        # which can be referenced as the unnecessary one if assigned to again before use.
+        # the third element is whether it has been seen in the function before.
+
+        for var in node.locals:  # initializes the values list with all the local variables.
+            values[var] = [0, None, False]
+
+        errors = []  # the list of problematic nodes
+        errorexists = False  # initially we assume there is no error
+
+        for block in node.nodes_of_class(astroid.ALL_NODE_CLASSES):
+            # this will go through all the nodes in the function and analyze them, based on print nodes.
+
+            if not isinstance(block.parent, astroid.FunctionDef):
+                # the reason we care if the parent is the FunctionDef is because the only Name
+                # node that this will be the case for is the Name node for the .return field of the FunctionDef
+                # which isn't technically part of the function, it just describes the return type, not value.
+
+                if isinstance(block, astroid.AssignName):
+                    # here we are checking if the node is one where assignment takes place.
+
+                    if block.name in values and values[block.name][0] > -1 and not isinstance(block.parent,
+                                                                                              astroid.AugAssign) and \
+                            values[block.name][2]:
+                        # if it is already in the values list and being reassigned before use then there
+                        # is unnecessary assignment and we should return True with the problematic node that
+                        # assigned before this assignment because it is useless.
+                        # KEY NOTE, this is unless it is part of an augmented assignment!
+                        errors.append(values[block.name][1].parent)  # this will add the parent node
+                        # so the whole line is highlighted
+                        errorexists = True
+
+                    elif block.name in values:
+                        # Either the name isn't in values or it is
+                        # in values and its value is -1 so it has been used.
+                        # so we just reset to now unused with the new value. Either way the same thing is done.
+                        values[block.name] = [0, block, True]
+
+                elif isinstance(block, astroid.Name) and block.name in values:
+                    # here we check if the value is being used.
+                    values[block.name] = [-1, None, True]
+
+        for val in values:
+            # this will check at the end if there were values unused.
+            if values[val][0] > -1 and values[val][2]:
+                errors.append(values[val][1].parent)
+                errorexists = True
+
+        return [errorexists, errors]
```

Comment:
The `errorexists` variable is unnecessary. Because you're now returning a *list*, returning an empty list is a perfectly valid case. This means that you can just return `errors`, and inside the main `visit` function, loop through `errors` to add a message for each one.

BTW one way you can tell this variable is unnecessary is that your code has the *invariant* that `errorexists` is True if and only if `errors` is non-empty.